molecule/src/molecule_auto_loader.cpp

```diff
@@ -290,24 +289,58 @@ void MoleculeAutoLoader::_loadMolecule(BaseMolecule& mol, bool query)
     {
         long long pos = _scanner->tell();
         _scanner->skipSpace();
-
         if (_scanner->lookNext() == '{')
         {
-            if (_scanner->findWord("molecule"))
+            if (_scanner->findWord("molecule")) // is it really reliable detection?
             {
+                using namespace rapidjson;
                 _scanner->seek(pos, SEEK_SET);
-                try
+//                try
                 {
-                    MoleculeJsonLoader loader(*_scanner);
-                    if (query)
-                        loader.loadQueryMolecule((QueryMolecule&)mol);
-                    else
-                        loader.loadMolecule((Molecule&)mol);
+                    Array<char> buf;
+                    _scanner->readAll(buf);
+                    buf.push(0);
+                    Document data;
+                    Value rgroups(kArrayType);
+                    Value mol_nodes(kArrayType);
+                    if ( data.Parse(buf.ptr()).HasParseError())
+                      throw Error("Error at parsing JSON: %s", buf.ptr());
+                    if( data.HasMember( "root" ) )
+                    {
+                        const Value& root = data["root"];
+                        const Value& nodes = root["nodes"];
+                        // rewind to first molecule node
+                        for( int i = 0; i < nodes.Size(); ++i )
+                        {
+                            const char* node_name = nodes[i]["$ref"].GetString();
+                            Value& node = data[ node_name ];
+                            std::string node_type = node["type"].GetString();
+                            if( node_type.compare("molecule") == 0 )
+                            {
+                                mol_nodes.PushBack( node, data.GetAllocator() );
+                            } else if (node_type.compare("rgroup") == 0 )
+                            {
+                                rgroups.PushBack( node, data.GetAllocator() );
+                            } else
```

Comment:
IMHO, this not a good way to parse JSON and then construct two nodes (`node` and `rgroup`). It can result in losing information from JSON. The ask of auto loader is just to identify the type and call a corresponding loader. The rest of parsing should be done inside the loader.

Reply:
Yeah, treat this as optional proposal.

Reply:
This is workaround was implemented for previous version of ket-format when it had multiple molecule nodes and it was impossible to handle it in one MoleculeJsonLoader class. I'd prefer to leave it as is until we have final ket-format version.